photon-ml/src/main/scala/com/linkedin/photon/ml/event/EventEmitter.scala

```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright 2016 LinkedIn Corp. All rights reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License. You may obtain a
+ * copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linkedin.photon.ml.event
+
+/**
+ * Base trait for all [[Event]] producers.
+ */
+trait EventEmitter {
+  private var registeredListeners = List[EventListener]()
+
+  /**
+   * Register an [[EventListener]] to consume [[Event]] objects produced by this class.
+   *
+   * @param listener The listener.
+   */
+  protected def registerListener(listener: EventListener): Unit = registeredListeners = listener :: registeredListeners
```

Comment:
This might belong to the category of issues we don't care about right now, but this guy is not thread-safe. The following sequence of events would break it:

Thread 1 calls `registerListener`, grabs `registeredListeners`, and then gets preempted.
Thread 2 calls `registerListener`, grabs `registeredListeners` (still the old one), and gets preempted.
Thread 1 resumes and finishes adding its listener
Thread 2 resumes and finishes

At this point the result from Thread 2 overwrites the result from Thread 1, and we lose that listener.